File: node/src/registers.ts
/**
 * ADXL355 register addresses and constants.
 * Based on ADXL354/ADXL355 Rev.D datasheet.
 */

/** Register addresses */
export const enum Reg {
  DEVID_AD = 0x00,
  DEVID_MST = 0x01,
  PARTID = 0x02,
  REVID = 0x03,
  STATUS = 0x04,
  FIFO_ENTRIES = 0x05,
  TEMP2 = 0x06,
  TEMP1 = 0x07,
  XDATA3 = 0x08,
  XDATA2 = 0x09,
  XDATA1 = 0x0a,
  YDATA3 = 0x0b,
  YDATA2 = 0x0c,
  YDATA1 = 0x0d,
  ZDATA3 = 0x0e,
  ZDATA2 = 0x0f,
  ZDATA1 = 0x10,
  FIFO_DATA = 0x11,
  OFFSET_X_H = 0x1e,
  OFFSET_X_L = 0x1f,
  OFFSET_Y_H = 0x20,
  OFFSET_Y_L = 0x21,
  OFFSET_Z_H = 0x22,
  OFFSET_Z_L = 0x23,
  ACT_EN = 0x24,
  ACT_THRESH_H = 0x25,
  ACT_THRESH_L = 0x26,
  ACT_COUNT = 0x27,
  FILTER = 0x28,
  FIFO_SAMPLES = 0x29,
  INT_MAP = 0x2a,
  SYNC = 0x2b,
  RANGE = 0x2c,
  POWER_CTL = 0x2d,
  SELF_TEST = 0x2e,
  RESET = 0x2f,
}

/** Expected device identity values (datasheet Rev.D, Tables 23-25) */
export const DEVID_AD_VALUE = 0xad;
export const DEVID_MST_VALUE = 0x1d;
export const PARTID_VALUE = 0xed;

/** Software reset code (datasheet Rev.D, Table 45) */
export const RESET_CODE = 0x52;

/** STATUS register bit positions (datasheet Rev.D, Table 27) */
export const STATUS_NVM_BUSY = 4;
export const STATUS_ACTIVITY = 3;
export const STATUS_FIFO_OVR = 2;
export const STATUS_FIFO_FULL = 1;
export const STATUS_DATA_RDY = 0;

/** FILTER register masks (datasheet Rev.D, Table 38) */
export const FILTER_ODR_MASK = 0x0f;
export const FILTER_ODR_SHIFT = 0;
export const FILTER_HPF_MASK = 0x70;
export const FILTER_HPF_SHIFT = 4;

/** RANGE register mask for bits 1:0 (datasheet Rev.D, Table 42) */
export const RANGE_SEL_MASK = 0x03;

/** SPI command helpers (datasheet Rev.D, SPI Protocol section) */
export function spiReadCmd(reg: number): number { return (reg << 1) | 0x01; }
export function spiWriteCmd(reg: number): number { return reg << 1; }

/** I2C addresses (datasheet Rev.D, Table 8) */
export const I2C_DEFAULT_ADDR = 0x1d;
export const I2C_ALTERNATE_ADDR = 0x53;

/** Acceleration range (datasheet Rev.D, Table 42) */
export enum Range {
  G2 = 0x01,
  G4 = 0x02,
  G8 = 0x03,
}

/** Power mode (datasheet Rev.D, Table 43: bit 0 = 1 => standby) */
export enum PowerMode {
  Standby = 1,
  Measurement = 0,
}

/** Output data rate */
export enum Odr {
  Hz4000 = 0,
  Hz2000 = 1,
  Hz1000 = 2,
  Hz500 = 3,
  Hz250 = 4,
  Hz125 = 5,
  Hz62_5 = 6,
  Hz31_25 = 7,
  Hz15_625 = 8,
  Hz7_813 = 9,
  Hz3_906 = 10,
}

/** Scaling constants (g per LSB) */
export const SCALE_2G_G_PER_LSB = 0.0000039;
export const SCALE_4G_G_PER_LSB = 0.0000078;
export const SCALE_8G_G_PER_LSB = 0.0000156;

/** Standard gravity (m/s²) */
export const STANDARD_GRAVITY_M_S2 = 9.80665;

File: rust/src/registers.rs
//! ADXL355 register addresses and bit masks.
//!
//! Preliminary values — verify against official ADXL355 datasheet.

/// Register addresses.
#[allow(non_snake_case, non_upper_case_globals)]
pub mod reg {
    /// Device ID (Analog Devices).
    pub const DEVID_AD: u8 = 0x00;
    /// MEMS sensor ID.
    pub const DEVID_MST: u8 = 0x01;
    /// Part ID.
    pub const PARTID: u8 = 0x02;
    /// Revision ID.
    pub const REVID: u8 = 0x03;
    /// Status register.
    pub const STATUS: u8 = 0x04;
    /// FIFO entry count.
    pub const FIFO_ENTRIES: u8 = 0x05;
    /// Temperature high byte.
    pub const TEMP2: u8 = 0x06;
    /// Temperature low byte.
    pub const TEMP1: u8 = 0x07;
    /// X-axis acceleration MSB.
    pub const XDATA3: u8 = 0x08;
    /// X-axis acceleration mid byte.
    pub const XDATA2: u8 = 0x09;
    /// X-axis acceleration LSB.
    pub const XDATA1: u8 = 0x0A;
    /// Y-axis acceleration MSB.
    pub const YDATA3: u8 = 0x0B;
    /// Y-axis acceleration mid byte.
    pub const YDATA2: u8 = 0x0C;
    /// Y-axis acceleration LSB.
    pub const YDATA1: u8 = 0x0D;
    /// Z-axis acceleration MSB.
    pub const ZDATA3: u8 = 0x0E;
    /// Z-axis acceleration mid byte.
    pub const ZDATA2: u8 = 0x0F;
    /// Z-axis acceleration LSB.
    pub const ZDATA1: u8 = 0x10;
    /// FIFO data read.
    pub const FIFO_DATA: u8 = 0x11;
    /// X-axis offset high byte.
    pub const OFFSET_X_H: u8 = 0x1E;
    /// X-axis offset low byte.
    pub const OFFSET_X_L: u8 = 0x1F;
    /// Y-axis offset high byte.
    pub const OFFSET_Y_H: u8 = 0x20;
    /// Y-axis offset low byte.
    pub const OFFSET_Y_L: u8 = 0x21;
    /// Z-axis offset high byte.
    pub const OFFSET_Z_H: u8 = 0x22;
    /// Z-axis offset low byte.
    pub const OFFSET_Z_L: u8 = 0x23;
    /// Activity detection enable.
    pub const ACT_EN: u8 = 0x24;
    /// Activity threshold high byte.
    pub const ACT_THRESH_H: u8 = 0x25;
    /// Activity threshold low byte.
    pub const ACT_THRESH_L: u8 = 0x26;
    /// Activity count.
    pub const ACT_COUNT: u8 = 0x27;
    /// Filter / output data rate control.
    pub const FILTER: u8 = 0x28;
    /// FIFO sample count threshold.
    pub const FIFO_SAMPLES: u8 = 0x29;
    /// Interrupt mapping.
    pub const INT_MAP: u8 = 0x2A;
    /// External synchronization control.
    pub const SYNC: u8 = 0x2B;
    /// Acceleration range selection.
    pub const RANGE: u8 = 0x2C;
    /// Power control.
    pub const POWER_CTL: u8 = 0x2D;
    /// Self-test control.
    pub const SELF_TEST: u8 = 0x2E;
    /// Software reset (write 0x52).
    pub const RESET: u8 = 0x2F;
}

/// Expected device identity values.
pub mod id {
    pub const DEVID_AD: u8 = 0xAD;
    pub const DEVID_MST: u8 = 0x1D;
    pub const PARTID: u8 = 0xED;
}

/// Software reset code.
pub const RESET_CODE: u8 = 0x52;

/// Power mode bit.
pub const POWER_MODE_BIT: u8 = 0;

/// Acceleration range selection.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Range {
    G2 = 0,
    G4 = 1,
    G8 = 2,
}

impl Range {
    /// Convert to register value.
    pub fn to_register(self) -> u8 {
        self as u8
    }

    /// Create from register value.
    pub fn from_register(val: u8) -> Option<Self> {
        match val & 0x03 {
            0 => Some(Range::G2),
            1 => Some(Range::G4),
            2 => Some(Range::G8),
            _ => None,
        }
    }

    /// Scale factor in g per LSB.
    pub fn scale_g_per_lsb(self) -> f32 {
        match self {
            Range::G2 => 0.0000039,
            Range::G4 => 0.0000078,
            Range::G8 => 0.0000156,
        }
    }
}

/// Power mode.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum PowerMode {
    Standby = 0,
    Measurement = 1,
}

/// Output data rate.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Odr {
    Hz4000 = 0,
    Hz2000 = 1,
    Hz1000 = 2,
    Hz500 = 3,
    Hz250 = 4,
    Hz125 = 5,
    Hz62_5 = 6,
    Hz31_25 = 7,
    Hz15_625 = 8,
    Hz7_813 = 9,
    Hz3_906 = 10,
}

/// Standard gravity (m/s²).
pub const STANDARD_GRAVITY_M_S2: f32 = 9.80665;
